src/Listener/JsonApiListener.php

```diff
@@ -0,0 +1,583 @@
+     * @return \Cake\ORM\AssociationCollection
+     */
+    protected function _stripNonContainedAssociations($table, $entity)
+    {
+        $associations = $table->associations();
+
+        foreach ($associations as $association) {
+            $associationKey = strtolower($association->name());
+            $entityKey = $association->table();
+
+            if (get_class($association) === 'Cake\ORM\Association\BelongsTo') {
+                $entityKey = Inflector::singularize($entityKey);
+            }
+
+            if (empty($entity->$entityKey)) {
+                $associations->remove($associationKey);
+            }
+        }
+
+        return $associations;
+    }
+
+    /**
+     * Returns a list with entity names of both passed entity name (as string)
+     * and all related/associated models. Used to read or generate NeoMerx
+     * schemas inside the view.
+     *
+     * @param string $entityName Name of the current entity
+     * @param \Cake\ORM\AssociationCollection $associations AssciationCollection
+     * @return array
+     */
+    protected function _getEntityList($entityName, $associations)
+    {
+        $entities = [$entityName];
+
+        foreach ($associations as $association) {
+            $entities[] = Inflector::singularize($association->name());
+        }
+
+        return $entities;
+    }
+
+    /**
+     * Generates a list for with all associated data (as produced by Containable
+     * and thus) present in the entity to be used for filling the top-level
+     * `included` node in the json response UNLESS user has specified listener
+     * config option 'include'.
+     *
+     * @param \Cake\ORM\AssociationCollection $associations AssociationCollection
+     * @return array
+     */
+    protected function _getIncludeList($associations)
+    {
+        if (!empty($this->config('include'))) {
+            return $this->config('include');
+        }
+
+        $result = [];
+        foreach ($associations as $association) {
+            if (is_a($association, '\Cake\ORM\Association\BelongsTo')) {
+                $include = Inflector::tableize($association->name());
+                $include = Inflector::singularize($include);
+
+                $result[] = $include;
+                continue;
+            }
+
+            // hasMany
+            $result[] = Inflector::tableize($association->name());
+        }
+
+        return $result;
+    }
+
+    /**
+     * Checks if data was posted to the Listener. If so then checks if the
+     * array (already converted from json) matches the expected JSON API
+     * structure for resources and if so, converts that array to CakePHP
+     * compatible format so it can be processed as usual from there.
+     *
+     * @return void
+     */
+    protected function _checkRequestData()
+    {
+        $requestMethod = $this->_controller()->request->method();
+
+        if ($requestMethod !== 'POST' && $requestMethod !== 'PATCH') {
```

Comment:
I tried both and only the && gave me the desired result (method skipped for everything but POST and PATCH)